### PR TITLE
doc for setting up Service Catalog with Prometheus metrics

### PR DESCRIPTION
**Pull request by jboyd01**
Documentation on how to view Service Catalog metrics and example YAML for deploying Prometheus to consume Catalog metrics.

### CODE REVIEW

File: contrib/examples/prometheus/prometheus.yml [hunk 1]
```diff
@@ -0,0 +1,341 @@
+#
+# Prometheus will run uner the "prometheus" Service Account, create
```
Comment (pmorie) [inline contrib/examples/prometheus/prometheus.yml]: typo: under

File: contrib/examples/prometheus/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+Service Catalog maintains metrics through the Prometheus client library and
+exposes them through the Prometheus http adapter at the /metrics endpoint.
+These metrics can be accessed directly via HTTP GET or more commonly scraped by
+the Prometheus monitoring application which persists metrics and facilitates
+analysis through a Web UI and powerful query language.
+
+Many metrics are not created and exposed until Service Catalog performs
+operations which would impact the metrics.  IE if you have no Service Brokers
+defined, there will be no metrics around class or plan count and likely no OSB
+Client operation metrics.  So before proceeding, its recommnded to you have a
```
Comment (pmorie) [inline contrib/examples/prometheus/README.md]: recommended

Comment (pmorie) [reply contrib/examples/prometheus/README.md]: also, "it's"

Comment (pmorie) [reply contrib/examples/prometheus/README.md]: sorry, mom was an english teacher

File: contrib/examples/prometheus/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# Prometheus Metrics
+
+Service Catalog maintains metrics through the Prometheus client library and
+exposes them through the Prometheus http adapter at the /metrics endpoint.
```
Comment (pmorie) [inline contrib/examples/prometheus/README.md]: This is the endpoint relative to the controller, right

File: contrib/examples/prometheus/prometheus.yml [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+# access to the necessary resources
```
Comment (MHBauer) [inline contrib/examples/prometheus/prometheus.yml]: What is "create access to the necessary resources"?

Comment (MHBauer) [reply contrib/examples/prometheus/prometheus.yml]: I think maybe period for both as separate sentences. Still not sure why we would need the second sentence. It's rbac, it's all about access.

File: contrib/examples/prometheus/README.md (hunk 3 above)
Comment (jboyd01) [reply contrib/examples/prometheus/README.md]: Yes, correct.  I've updated this doc to be explicit.

File: contrib/examples/prometheus/prometheus.yml [hunk 5]
```diff
@@ -0,0 +1,341 @@
+#
+# Prometheus will run under the "prometheus" Service Account, create
```
Comment (MHBauer) [inline contrib/examples/prometheus/prometheus.yml]: period at the comma. remove the rest.

File: contrib/examples/prometheus/prometheus.yml (hunk 1 above)
Comment (jboyd01) [reply contrib/examples/prometheus/prometheus.yml]: Thanks for the review @pmorie, typos matter.  All have bee addressed.

File: contrib/examples/prometheus/README.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# Prometheus Metrics
+
+Service Catalog maintains metrics through the Prometheus client library and
+exposes them through the Prometheus http adapter at the Controller's /metrics
+endpoint. These metrics can be accessed directly via HTTP GET or more commonly
+scraped by the Prometheus monitoring application which persists metrics and
+facilitates analysis through a Web UI and powerful query language.
+
+Many metrics are not created and exposed until Service Catalog performs
+operations which would impact the metrics.  IE if you have no Service Brokers
```
Comment (MHBauer) [inline contrib/examples/prometheus/README.md]: IE -> In other words,

File: contrib/examples/prometheus/README.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+defined, there will be no metrics around class or plan count and likely no OSB
+Client operation metrics.  So before proceeding, it's recommended you have a
+Broker defined and created a Service Instance.
```
Comment (MHBauer) [inline contrib/examples/prometheus/README.md]: Not being familiar with prometheus, what does 'no metrics' look like? Does it mean everything is 0 or that until we do this prometheus won't even show what we've defined in the svccat?

File: contrib/examples/prometheus/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+To view the raw metrics:
+
+```
+# setup a port forward so we can curl against Controller Manager
+
+$ kubectl get pods -l app=catalog-catalog-controller-manager -n catalog -o name | \
+    sed 's/^.*\///' | xargs -I{} kubectl port-forward {} -n catalog 8089:8080 &
+
+$ curl -s http://localhost:8089/metrics  | grep servicecatalog
```
Comment (MHBauer) [inline contrib/examples/prometheus/README.md]: this drops us into a pseudo shell of some sort?

File: contrib/examples/prometheus/README.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+preconfigured to gather Kubernetes platform and node metrics.  If you deploy the
+Service Catalog Controller Manager via Helm with the optional
+`enablePrometheusScrape` parameter set to true (see
+[charts/catalog/values.yaml](../../../charts/catalog/values.yaml)), this
+configuration will direct Prometheus to automatically scrape custom metrics
+exposed from Service Catalog as well.  Most any Prometheus configuration for
+Kubernetes (ie [Prometheus
+Operator](https://github.com/coreos/prometheus-operator)) will pick up the
+Service Catalog metrics as long as it's looking for pods with the
+`prometheus.io/scrape` annotation.
```
Comment (MHBauer) [inline contrib/examples/prometheus/README.md]: this is really neat!

File: contrib/examples/prometheus/README.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+servicecatalog_osb_request_count{broker="ups-broker",method="Bind",status="2xx"} 41
+servicecatalog_osb_request_count{broker="ups-broker",method="GetCatalog",status="2xx"} 1
+servicecatalog_osb_request_count{broker="ups-broker",method="ProvisionInstance",status="2xx"} 2
+```
+
+Alternatively, and the more common approach to utlizing metrics, deploy
+Prometheus.  [This YAML](prometheus.yml) creates a Prometheus instance
+preconfigured to gather Kubernetes platform and node metrics.  If you deploy the
+Service Catalog Controller Manager via Helm with the optional
+`enablePrometheusScrape` parameter set to true (see
```
Comment (MHBauer) [inline contrib/examples/prometheus/README.md]: We should probably write the exact `helm set` here in addition to the general descriptive guideline.

File: contrib/examples/prometheus/README.md (hunk 8 above)
Comment (jboyd01) [reply contrib/examples/prometheus/README.md]: The first command actually backgrounds a port forward allowing you to hit http://localhost:8089 and kube will forward the request to the catalog-controller pod's end point.

**Comment by jboyd01** [comment]
Thanks for the review @MHBauer.  I fixed the over zealous comments in the YAML file and addressed the other issues.  I meant this more as a developer/experimenter guide with the plan of adding real/better metrics documentation.  Once we have more Catalog metrics I hope we can call out specific items to monitor in production and perhaps even suggestions on what to alert off from.

**Comment by pmorie** [comment]
Thanks for this, @jboyd01!

**Pull request merged**